Bug: Auto Fix comma-dangle wipes file with 'always' and a final paren wrapped item with no comma

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**What version of ESLint are you using?**
2.10.2

**What parser (default, Babel-ESLint, etc.) are you using?**
None

**Please show your full configuration:**
```json
{
  "rules": {
    "comma-dangle": ["error", "always"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
I ran `eslint --fix commas.js`

**`commas.js`:**
```js
module.exports = [
'good',
('bad')
];
```

**What did you expect to happen?**
An additional training comma added: `('bad'),`

**What actually happened? Please include the actual, raw output from ESLint.**
I got an empty file in commas.js. I've reduced the test case to a minimum, but this was happening in a large .jsx file, which also ended up empty.

Here's the `--debug` output:
```
  eslint:cli Running on files +0ms
  eslint:ignored-paths Looking for ignore file in d:\git\eslint-bug +49ms
  eslint:ignored-paths Could not find ignore file in cwd +2ms
  eslint:glob-util Creating list of files to process. +1ms
  eslint:cli-engine Processing d:\git\eslint-bug\commas.js +4ms
  eslint:cli-engine Linting d:\git\eslint-bug\commas.js +2ms
  eslint:config Constructing config for d:\git\eslint-bug\commas.js +1ms
  eslint:config Using .eslintrc and package.json files +2ms
  eslint:config Loading d:\git\eslint-bug\.eslintrc +3ms
  eslint:config-file Loading config file: d:\git\eslint-bug\.eslintrc +2ms
  eslint:config Using d:\git\eslint-bug\.eslintrc +77ms
  eslint:config Merging command line environment settings +1ms
  eslint:config-ops Apply environment settings to config +1ms
  eslint:cli-engine Linting code for d:\git\eslint-bug\commas.js (pass 1) +2ms
  eslint:cli-engine Generating fixed text for d:\git\eslint-bug\commas.js (pass 1) +50ms
  eslint:text-fixer Applying fixes +2ms
  eslint:text-fixer Found fixes to apply +2ms
  eslint:cli-engine Linting code for d:\git\eslint-bug\commas.js (pass 2) +2ms
  eslint:cli-engine Generating fixed text for d:\git\eslint-bug\commas.js (pass 2) +3ms
  eslint:text-fixer Applying fixes +2ms
  eslint:text-fixer No source code to fix +2ms
  eslint:cli-engine Linting complete in: 165ms +2ms
  eslint:cli Fix mode enabled - applying fixes +1ms

d:\git\eslint-bug\commas.js
  3:8  error  Parsing error: Unexpected token )

✖ 1 problem (1 error, 0 warnings)
```

Note the *No source code to fix* part after the 2nd pass

The parens around the final item are key - removing them makes everything work again.

--- a/lib/rules/comma-dangle.js
+++ b/lib/rules/comma-dangle.js
@@ -143,13 +143,13 @@ module.exports = {
             }
 
             var sourceCode = context.getSourceCode(),
-                trailingToken;
-
-            // last item can be surrounded by parentheses for object and array literals
-            if (node.type === "ObjectExpression" || node.type === "ArrayExpression") {
-                trailingToken = sourceCode.getTokenBefore(sourceCode.getLastToken(node));
-            } else {
+                penultimateToken = lastItem,
                 trailingToken = sourceCode.getTokenAfter(lastItem);
+
+            // Skip close parentheses.
+            while (trailingToken.value === ")") {
+                penultimateToken = trailingToken;
+                trailingToken = sourceCode.getTokenAfter(trailingToken);
             }
 
             if (trailingToken.value !== ",") {
@@ -158,7 +158,7 @@ module.exports = {
                     loc: lastItem.loc.end,
                     message: MISSING_MESSAGE,
                     fix: function(fixer) {
-                        return fixer.insertTextAfter(lastItem, ",");
+                        return fixer.insertTextAfter(penultimateToken, ",");
                     }
                 });
             }

--- a/tests/lib/rules/comma-dangle.js
+++ b/tests/lib/rules/comma-dangle.js
@@ -846,6 +846,26 @@ ruleTester.run("comma-dangle", rule, {
             parserOptions: { sourceType: "module" },
             options: ["always-multiline"],
             errors: [{message: "Missing trailing comma.", type: "ExportSpecifier"}]
+        },
+
+        // https://github.com/eslint/eslint/issues/6233
+        {
+            code: "var foo = {a: (1)}",
+            output: "var foo = {a: (1),}",
+            options: ["always"],
+            errors: [{message: "Missing trailing comma.", type: "Property"}]
+        },
+        {
+            code: "var foo = [(1)]",
+            output: "var foo = [(1),]",
+            options: ["always"],
+            errors: [{message: "Missing trailing comma.", type: "Literal"}]
+        },
+        {
+            code: "var foo = [\n1,\n(2)\n]",
+            output: "var foo = [\n1,\n(2),\n]",
+            options: ["always-multiline"],
+            errors: [{message: "Missing trailing comma.", type: "Literal"}]
         }
     ]
 });

